### ISSUE TITLE
Plugin can crash Joplin and make it unusuable

### ISSUE THREAD

**Issue by pvojnisek** [laurent22/joplin#4727]
Plugins can crash Joplin and make it unusable. It can be reproducible following the steps bellow.

1. Install PlantUML plugin: https://github.com/uphy/joplin-plantuml-plugin
2. Paste an example PlantUML code into a blank new notebook (from here: https://plantuml.com/salt )
3. The editor window crashes
4. The same crash happens after restarting Joplin
5. Joplin can not be started anymore

The example PlantUML:
```
@startsalt
{+
{* File | Edit | Source | Refactor }
{/ General | Fullscreen | Behavior | Saving }
{
{ Open image in: | ^Smart Mode^ }
[X] Smooth images when zoomed
[X] Confirm image deletion
[ ] Show hidden images
}
[Close]
}
@endsalt
```

Joplin sends this error message:

```
Error
Joplin encountered a fatal error and could not continue. To report the error, please copy the *entire content* of this page and post it on Joplin forum or GitHub.

Message
Version info
Joplin for Desktop

Copyright © 2016-2021 Laurent Cozic
Joplin 1.7.11 (prod, linux)

Client ID: 17c149a0a4094fbea7714d83bbecd8ad
Sync Version: 2
Profile Version: 34
Keychain Supported: No
Component stack

    in NoteEditor (created by Connect(NoteEditor))
    in Connect(NoteEditor) (created by ResizableLayout)
    in div (created by styled.div)
    in styled.div (created by ResizableLayout)
    in div (created by ResizableLayout)
    in div (created by ResizableLayout)
    in ResizableLayout (created by MainScreenComponent)
    in div (created by MainScreenComponent)
    in MainScreenComponent (created by Connect(MainScreenComponent))
    in Connect(MainScreenComponent) (created by NavigatorComponent)
    in div (created by NavigatorComponent)
    in NavigatorComponent (created by Connect(NavigatorComponent))
    in Connect(NavigatorComponent) (created by RootComponent)
    in ThemeProvider (created by RootComponent)
    in StyleSheetManager (created by RootComponent)
    in RootComponent (created by Connect(RootComponent))
    in Connect(RootComponent)
    in ErrorBoundary
    in Provider
```

**Comment by pvojnisek** [comment]
I've added the same error report to the plugin repo: https://github.com/uphy/joplin-plantuml-plugin/issues/3

**Comment by roman-r-m** [comment]
I also saw this in dev console:
```
Uncaught Unknown state undefined
token @ /home/roman/.config/joplindev-desktop/cache/com.github.uphy.PlantUmlPlugin/codeMirrorPlantUmlPlugin.js:1
token @ /home/roman/dev/joplin/packages/app-desktop/node_modules/codemirror/addon/mode/multiplex.js:84
readToken @ /home/roman/dev/joplin/packages/app-desktop/node_modules/codemirror/lib/codemirror.js:1168
runMode @ /home/roman/dev/joplin/packages/app-desktop/node_modules/codemirror/lib/codemirror.js:1225
highlightLine @ /home/roman/dev/joplin/packages/app-desktop/node_modules/codemirror/lib/codemirror.js:1071
getLineStyles @ /home/roman/dev/joplin/packages/app-desktop/node_modules/codemirror/lib/codemirror.js:1115
buildLineContent @ /home/roman/dev/joplin/packages/app-desktop/node_modules/codemirror/lib/codemirror.js:1748
getLineContent @ /home/roman/dev/joplin/packages/app-desktop/node_modules/codemirror/lib/codemirror.js:2138
buildLineElement @ /home/roman/dev/joplin/packages/app-desktop/node_modules/codemirror/lib/codemirror.js:2219
patchDisplay @ /home/roman/dev/joplin/packages/app-desktop/node_modules/codemirror/lib/codemirror.js:4208
updateDisplayIfNeeded @ /home/roman/dev/joplin/packages/app-desktop/node_modules/codemirror/lib/codemirror.js:4115
endOperation_W1 @ /home/roman/dev/joplin/packages/app-desktop/node_modules/codemirror/lib/codemirror.js:3824
endOperations @ /home/roman/dev/joplin/packages/app-desktop/node_modules/codemirror/lib/codemirror.js:3801
(anonymous) @ /home/roman/dev/joplin/packages/app-desktop/node_modules/codemirror/lib/codemirror.js:3790
finishOperation @ /home/roman/dev/joplin/packages/app-desktop/node_modules/codemirror/lib/codemirror.js:2050
endOperation @ /home/roman/dev/joplin/packages/app-desktop/node_modules/codemirror/lib/codemirror.js:3787
runInOp @ /home/roman/dev/joplin/packages/app-desktop/node_modules/codemirror/lib/codemirror.js:3916
operation @ /home/roman/dev/joplin/packages/app-desktop/node_modules/codemirror/lib/codemirror.js:8590
forEachCodeMirror @ /home/roman/dev/joplin/packages/app-desktop/node_modules/codemirror/lib/codemirror.js:6605
(anonymous) @ /home/roman/dev/joplin/packages/app-desktop/node_modules/codemirror/lib/codemirror.js:6622
setTimeout (async)
(anonymous) @ /home/roman/dev/joplin/packages/app-desktop/node_modules/codemirror/lib/codemirror.js:6620
```

**Comment by laurent22** [comment]
There's two problem here:

1. That the plugin crashes the app. For this, on the app side there should be some way to make sure that a content script cannot do that.

2. That it's not possible to recover from it by restarting the app. I guess for this we should have some kind of safe mode, with plugins and possibly Markdown rendering disabled.

**Comment by Ahmad45123** [comment]
I wasn't able to replicate the crash on my end for some reason.. I'm not sure but the plugin didn't even work.
![image](https://user-images.githubusercontent.com/8184424/112225607-b6070200-8c35-11eb-96f3-58c90d804ca7.png)
It doesn't render and there's no errors or anything though the plugin did load correctly according to console. Hmm.

> That it's not possible to recover from it by restarting the app. I guess for this we should have some kind of safe mode, with plugins and possibly Markdown rendering disabled.

How do you suggest for this to be implemented ? Maybe just a command line argument ? `--safemode` ?
It'd disable plugin loading and markdown rendering as you suggested and also append "[ SAFE MODE ]" or something to the title bar.

**Comment by pvojnisek** [comment]
@Ahmad45123 
> I wasn't able to replicate the crash on my end for some reason.. I'm not sure but the plugin didn't even work.

Try to insert code block characters before and after like this:
```plantuml

```

**Comment by pvojnisek** [comment]
@laurent22 
Can you recommend a way to disable the plugin and start Joplin?

**Comment by roman-r-m** [comment]
Delete the `plugins` folder in our profile, or just the PlantUML plugin file.

**Comment by mablin7** [comment]
Strange, when I first paste the code into the markdown editor, the diagram shows up correctly, and the error message is logged in the console but no crash. If I then switch to rich text editor, there's still no issue. It is only when I switch back to the markdown editor that the crash occurs.

**Comment by roman-r-m** [comment]
Yes, noticed it too. But then it'll crash immediately on next startup

**Comment by mablin7** [comment]
Yes, same for me

**Comment by mablin7** [comment]
I submitted a PR to the plugin fixing this, so it should be usable once it's merged.



> Strange, when I first paste the code into the markdown editor, the diagram shows up correctly, and the error message is logged in the console but no crash. If I then switch to rich text editor, there's still no issue. It is only when I switch back to the markdown editor that the crash occurs.

I found out that this happens because when typing/pasting in code into an already open editor, CodeMirror is called from an event handler, so the exception doesn't propagate through the component tree and React keeps rendering as usual, but the editor itself will of course be broken. But when the editor is mounted, CodeMirror and thus the plugin, is called from the component's render function, from where it can wreck the whole component tree.

There's some discussion on the forum about the issue of plugins crashing Joplin on this thread:
https://discourse.joplinapp.org/t/improving-plugin-security/16163
